File: merge_remaining_branches.py
#!/usr/bin/env python3
"""
Merge Remaining Cursor Branches - Simple merge script
"""

import subprocess
import sys
from datetime import datetime

def run_command(cmd, capture_output=True, check=False):
    """Run a shell command and return the result"""
    try:
        print(f"→ {cmd}")
        result = subprocess.run(
            cmd,
            shell=True,
            capture_output=capture_output,
            text=True,
            check=check
        )
        if result.stdout and capture_output:
            print(result.stdout)
        if result.stderr and result.returncode != 0:
            print(f"Error: {result.stderr}")
        return result
    except subprocess.CalledProcessError as e:
        print(f"⚠ Command failed with return code {e.returncode}")
        return e
    except Exception as e:
        print(f"⚠ Exception: {e}")
        return None

def get_remaining_branches():
    """Get all remaining cursor branches"""
    result = run_command("git branch -r | grep 'cursor/fix-errors-and-merge-to-main' | head -50")
    if not result or result.returncode != 0:
        return []
    
    branches = []
    for line in result.stdout.strip().split('\n'):
        if line.strip():
            branch = line.strip().replace('origin/', '')
            branches.append(branch)
    
    return branches

def merge_branch(branch_name):
    """Attempt to merge a branch into main"""
    print(f"\n🔄 Merging {branch_name}...")
    
    # Try to merge
    result = run_command(f"git merge origin/{branch_name} --no-edit -m 'Merge branch {branch_name}'")
    
    if result and result.returncode == 0:
        print(f"✅ Successfully merged {branch_name}")
        return True
    else:
        print(f"❌ Failed to merge {branch_name}")
        # Abort the merge
        run_command("git merge --abort")
        return False

def main():
    """Main execution function"""
    print(f"🚀 Starting Remaining Branches Merge at {datetime.now()}")
    print(f"{'='*60}\n")
    
    # Get remaining branches
    branches = get_remaining_branches()
    if not branches:
        print("❌ No remaining branches found")
        return 1
    
    print(f"Found {len(branches)} branch(es) to merge:")
    for branch in branches:
        print(f"  • {branch}")
    print()
    
    # Track results
    successful_merges = []
    failed_merges = []
    
    # Process each branch
    for i, branch in enumerate(branches, 1):
        print(f"\n[{i}/{len(branches)}] Processing {branch}...")
        
        success = merge_branch(branch)
        
        if success:
            successful_merges.append(branch)
        else:
            failed_merges.append(branch)
    
    # Final summary
    print(f"\n{'='*60}")
    print("📊 MERGE SUMMARY")
    print(f"{'='*60}")
    print(f"Total branches: {len(branches)}")
    print(f"✅ Successfully merged: {len(successful_merges)}")
    print(f"❌ Failed to merge: {len(failed_merges)}")
    
    # Push changes
    if successful_merges:
        print(f"\n📤 Pushing merged changes...")
        push_result = run_command("git push origin main")
        if push_result and push_result.returncode == 0:
            print("✅ Successfully pushed all changes!")
        else:
            print("❌ Failed to push changes")
    
    print(f"\n🏁 Merge process completed at {datetime.now()}")
    
    return 0 if not failed_merges else 1

if __name__ == "__main__":
    try:
        sys.exit(main())
    except KeyboardInterrupt:
        print("\n\n⚠️  Interrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n\n💥 Fatal error: {e}")
        sys.exit(1)

File: merge_remaining_branches.js
#!/usr/bin/env node
import { execSync } from 'child_process';

function runCommand(command, description) {
  try {
    console.log(`🔄 ${description}...`);
    const result = execSync(command, { 
      cwd: '/workspace', 
      encoding: 'utf8',
      stdio: 'pipe'
    });
    console.log(`✅ ${description} completed`);
    return { success: true, output: result };
  } catch (error) {
    console.log(`❌ ${description} failed: ${error.message}`);
    return { success: false, error: error.message, output: error.stdout || error.stderr };
  }
}

async function mergeBranch(branchName) {
  console.log(`\n🔄 Processing branch: ${branchName}`);
  
  // Check if branch exists
  const branchCheck = runCommand(`git show-ref --verify --quiet refs/remotes/origin/${branchName}`, `Checking if branch ${branchName} exists`);
  if (!branchCheck.success) {
    console.log(`⏭️  Branch ${branchName} doesn't exist, skipping`);
    return false;
  }

  // Try to merge the branch
  const mergeResult = runCommand(`git merge origin/${branchName} --no-edit`, `Merging ${branchName}`);
  
  if (mergeResult.success) {
    console.log(`✅ Successfully merged ${branchName}`);
    return true;
  } else {
    // Check if it's a conflict
    if (mergeResult.output.includes('CONFLICT') || mergeResult.error.includes('CONFLICT')) {
      console.log(`⚠️  Merge conflict in ${branchName}, attempting to resolve...`);
      
      // Check git status to see what conflicts we have
      const statusResult = runCommand(`git status --porcelain`, `Checking git status for conflicts`);
      console.log(`Git status: ${statusResult.output}`);
      
      // Try to resolve conflicts automatically
      const addResult = runCommand(`git add .`, `Adding all changes`);
      
      if (addResult.success) {
        const commitResult = runCommand(`git commit --no-edit`, `Committing resolved conflicts`);
        if (commitResult.success) {
          console.log(`✅ Successfully resolved conflicts for ${branchName}`);
          return true;
        } else {
          console.log(`❌ Failed to commit resolved conflicts for ${branchName}`);
          return false;
        }
      } else {
        console.log(`❌ Failed to add changes for ${branchName}`);
        return false;
      }
    } else {
      console.log(`❌ Failed to merge ${branchName}: ${mergeResult.error}`);
      return false;
    }
  }
}

async function main() {
  try {
    console.log('🚀 Starting remaining branch merge process...');
    
    // Ensure we're on main and up to date
    runCommand('git checkout main', 'Switching to main branch');
    runCommand('git pull origin main', 'Pulling latest main');
    
    // Get list of remaining fix-errors branches
    const branchList = [
      'cursor/fix-errors-and-merge-to-main-0daa', 
      'cursor/fix-errors-and-merge-to-main-1182',
      'cursor/fix-errors-and-merge-to-main-127b',
      'cursor/fix-errors-and-merge-to-main-29a0',
      'cursor/fix-errors-and-merge-to-main-3f2f',
      'cursor/fix-errors-and-merge-to-main-5588',
      'cursor/fix-errors-and-merge-to-main-637b',
      'cursor/fix-errors-and-merge-to-main-654f',
      'cursor/fix-errors-and-merge-to-main-778a',
      'cursor/fix-errors-and-merge-to-main-bf77',
      'cursor/fix-errors-and-merge-to-main-c6fa',
      'cursor/fix-errors-and-merge-to-main-c928',
      'cursor/fix-errors-and-merge-to-main-e1e0',
      'fix-errors-final',
      'fix-typescript-errors-main'
    ];

    let successCount = 0;
    let conflictCount = 0;
    let errorCount = 0;

    for (const branch of branchList) {
      const result = await mergeBranch(branch);
      if (result === true) {
        successCount++;
      } else if (result === false) {
        conflictCount++;
      } else {
        errorCount++;
      }
    }

    console.log('\n📊 Merge Summary:');
    console.log(`  ✅ Successfully merged: ${successCount}`);
    console.log(`  ⚠️  Conflicts/Issues: ${conflictCount}`);
    console.log(`  ❌ Errors: ${errorCount}`);

    if (successCount > 0) {
      console.log('\n🚀 Pushing merged changes...');
      const pushResult = runCommand('git push origin main', 'Pushing merged changes to main');
      
      if (pushResult.success) {
        console.log('✅ Successfully pushed all merged changes to main');
      } else {
        console.log('❌ Failed to push changes:', pushResult.error);
      }
    }

  } catch (error) {
    console.error('❌ Error in main process:', error.message);
  }
}

main().catch(console.error);